wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -198,32 +201,42 @@ applyBlock pw@PassiveWallet{..} b = do
               restoreKnownWallet pw rootId
 
           -- Beginning with the oldest missing block, update each lagging account.
-          let applyOne (block, toAccts) = runExceptT (applyOneBlock k (Just toAccts) block)
+          let applyOne (block, toAccts) = runExceptT (applyOneBlock k withinK (Just toAccts) block)
           case toApply of
               []       -> return () -- nothing to do!
               (bk:bks) -> do
                   failures <- mapM applyOne (getOldestFirst $ gatherAcctsPerBlock (bk :| bks))
 
                   case NEM.fromMap . Map.unions . map NEM.toMap . lefts $ failures of
-                      Nothing       -> return ()                         -- OK, no failures, we are done!
-                      Just moreErrs -> handleApplyBlockErrors k moreErrs -- Try again, better luck next time.
+                      Nothing       -> return ()                                 -- OK, no failures, we are done!
+                      Just moreErrs -> handleApplyBlockErrors k withinK moreErrs -- Try again, better luck next time.
 
       -- Try to apply a single block, failing if it does not fit onto the most recent checkpoint.
       applyOneBlock :: Node.SecurityParameter
+                    -> Bool
                     -> Maybe (Set HdAccountId)
                     -> ResolvedBlock
                     -> ExceptT (NonEmptyMap HdAccountId ApplyBlockFailed) IO ()
-      applyOneBlock k accts b' = ExceptT $ do
+      applyOneBlock k withinK accts b' = ExceptT $ do
           ((ctxt, blocksByAccount), metas) <- prefilterBlock' pw b'
           -- apply block to all Accounts in all Wallets
-          mConfirmed <- update' _wallets $ ApplyBlock k ctxt accts blocksByAccount
-          case mConfirmed of
-              Left  errs      -> return (Left errs)
-              Right confirmed -> do
-                  modifyMVar_ _walletSubmission (return . Submission.remPending confirmed)
-                  mapM_ (putTxMeta _walletMeta) metas
-                  createCheckpointIfNeeded
-                  return $ Right ()
+          case Map.null blocksByAccount of
+               -- If this 'ResolvedBlock' is not relevant to us @and@ we are
+               -- very far from the tip (more than @k@ blocks) there is no
+               -- point applying this block, even just for the checkpoint, as
+               -- for rollback we can use sparse checkpoints anyway.
+               True | not withinK -> pure $ Right ()
```

Comment:
We reviewed this with @matt-noonan and we found at least one "gap" (no pun intended): it might happen that you do an initial full sync with the blockchain, and now you go from "no checkpoints" to "k checkpoints" once you start to approach the tip.

But now, if you shutdown the wallet and you wait more than `k` blocks to re-open it, it means that now you will _not_ apply blocks immediately, but you will wait till you are within `k` _again_. This means you are creating such sparse checkpoints:

```
o--- none -- | k --- | --- none --- | k ...
```

And this means that you have like "Clusters" of dense checkpoints alternating clusters of "no checkpoints", which is going to be problematic (probably).